enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -229,35 +230,101 @@ doing all of your testing in a virtual network built on top of it.
 [metallb-operator](https://github.com/cybertron/metallb-operator) available and
 a [video demo](https://www.youtube.com/watch?v=WgOZno0D7nw).
 
-#### Alternative Integration: Cloud Controller Manager
-
-An alternative integration approach would be via a cloud controller manager
-(CCM). An example of this is the [packet.net
-CCM](https://github.com/packethost/packet-ccm), which ensures MetalLB is
-deployed and also configures it properly to work in packet.net’s BGP
-environment.
-
-These integration options must be explored in more detail as part of a more
-detailed integration proposal.
+#### Configuration Interfaces
+
+MetalLB's current configuration interface is a single `ConfigMap`.  We only
+want to support CRD based configuration through OpenShift.  The ultimate goal
+will be to enhance MetalLB with a native CRD based configuration interface.
+The design work has begun here:
+
+https://github.com/metallb/metallb/pull/833
+
+Since this is an upstream effort, the completion of this work may take longer
+than the timeline for when we want to ship the first piece of MetalLB support
+for OpenShift, which will be to support MetalLB in layer2 mode only.
+
+To mitigate this timing risk, we have another option for a way to configure the
+layer2 mode of MetalLB using an existing OpenShift configuration option.  As a
+backup plan, we can have the OpenShift version of the MetalLB operator build on
+top of the existing feature to automatically assign external IPs to services of
+`type=LoadBalancer`.  For example, take this sample configuration:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Network
+metadata:
+  name: cluster
+spec:
+  ...
+  externalIP:
+    autoAssignCIDRs:
+    - 10.10.10.0/24
+```
+
+This makes OpenShift automatically set the `externalIP` field, but nothing will
+make sure those addresses are reachable automatically.  That's where MetalLB
+comes in.  If the OpenShift MetalLB operator is enabled, the operator can
+automatically generate a MetalLB config map based on this cluster
+configuration.  The operator will read the cluster network configuration and
+generate config like this:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: metallb-system
+  name: config
+data:
+  config: |
+    address-pools:
+    - name: openshift
+      protocol: layer2
+      auto-assign: false
+      addresses:
+      - 10.10.10.0/24
+```
+
+Note how the configuration copies over the CIDR for the externalIP allocator
+used in OpenShift.  This makes MetalLB expect to see addresses in this range
+used for load balancer Services.  Also note the use of `auto-assign: false`.
+In this scenario, we're letting OpenShift do the address allocation piece.
+
+If we support this as a configuration interface, customers will be able to opt
+for the more feature rich CRD interface offered by MetalLB natively
+once it is ready.  As long as the `AutoAssignCIDRs` feature is not enabled on
+the OpenShift side, there should not be a conflict.
```

Comment:
how about abnormal cases where u have Metallb then someone config AutoAssignCIDRs , or the other way around do we need to detect and fail the config if the field is in use ?

Reply:
AutoAssignCIDRs is basically an IPAM method for service external IPs? So in this case, we would be giving the option to use this IPAM instead of the MetalLB one?

Whats the advantage of this over a custom CRD (that may not be the final one)? Users will still need to migrate from this configuration method to the new, final one?

Reply:
AutoAssignCIDRs is basically an IPAM method for service external IPs? So in this case, we would be giving the option to use this IPAM instead of the MetalLB one?

Whats the advantage of this over a custom CRD (that may not be the final one)? Users will still need to migrate from this configuration method to the new, final one?

Reply:
> AutoAssignCIDRs is basically an IPAM method for service external IPs? So in this case, we would be giving the option to use this IPAM instead of the MetalLB one?

Yes, that's right.

> Whats the advantage of this over a custom CRD (that may not be the final one)? Users will still need to migrate from this configuration method to the new, final one?

Advantages:

- no new CRD required.  It's a simple an existing configuration interface that should be able to support indefinitely (for a limited layer2 only use case anyway).  So I think this is a path that won't *require* any migration, unless a customer chooses to for more config capabilities.

Disadvantages:

- Doesn't support all of the config flexibliity for layer2 that metallb is capable of (would come later with our full CRD interface)